Managers/UTMLogging.m

```diff
@@ -0,0 +1,215 @@
+//
+// Copyright © 2020 osy. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+#import <pthread.h>
+#import <stdio.h>
+#import <unistd.h>
+#import "UTMLogging.h"
+
+static const int kLogBufferSize = 4096;
+static UTMLogging *gLoggingInstance;
+
+@interface UTMLogging ()
+
+- (void)didRecieveNewLine:(NSString *)line onDescriptor:(int)fd;
+
+@end
+
+@implementation UTMLogging {
+    pthread_t _stdout_thread;
+    pthread_t _stderr_thread;
+    NSOutputStream *_stream;
+    int _real_stdout;
+    int _real_stderr;
+    int _stdout;
+    int _stderr;
+}
+
+void *utm_logging_thread_stdout(void *arg) {
+    UTMLogging *self = (__bridge_transfer UTMLogging *)arg;
+    char ch;
+    NSMutableData *data = [NSMutableData dataWithCapacity:kLogBufferSize];
+    while (read(self->_stdout, &ch, 1) > 0) {
+        [data appendBytes:&ch length:1];
+        if (ch == '\n') {
+            NSString *line = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
+            data = [NSMutableData dataWithCapacity:kLogBufferSize];
+            [self didRecieveNewLine:line onDescriptor:self->_stdout];
+        }
+    }
+    return NULL;
+}
+
+void *utm_logging_thread_stderr(void *arg) {
+    UTMLogging *self = (__bridge_transfer UTMLogging *)arg;
+    char ch;
+    NSMutableData *data = [NSMutableData dataWithCapacity:kLogBufferSize];
+    while (read(self->_stderr, &ch, 1) > 0) {
+        [data appendBytes:&ch length:1];
+        if (ch == '\n') {
+            NSString *line = [[NSString alloc] initWithData:data encoding:NSASCIIStringEncoding];
+            data = [NSMutableData dataWithCapacity:kLogBufferSize];
+            [self didRecieveNewLine:line onDescriptor:self->_stderr];
+        }
+    }
+    return NULL;
+}
+
++ (void)initialize {
+    static BOOL initialized = NO;
+    if (!initialized) {
+        initialized = YES;
+        gLoggingInstance = [[UTMLogging alloc] init];
+    }
+}
+
++ (UTMLogging *)sharedInstance {
+    return gLoggingInstance;
+}
+
+- (id)init {
+    self = [super init];
+    if (self) {
+        if (gLoggingInstance != nil) {
+            NSLog(@"Trying to init more than one instance of UTMLogging!");
+            return nil;
+        }
+        int success = 0;
+        int new_stdout[2];
+        int new_stderr[2];
+        _stdout = _stderr = -1;
+        _real_stdout = _real_stderr = -1;
+        new_stdout[0] = new_stdout[1] = -1;
+        new_stderr[0] = new_stderr[1] = -1;
```

Comment:
```
int new_stdout[] = {-1, -1};
int new_stderr[] = {-1, -1};
```?

Reply:
Oops sorry didn't see this comment! I thought of that but decided it's more consistent with the other lines of code to have it this way. It clarifies the intent that those different variables are related to each other.

Reply:
Fair enough. I think you actually committed first, and I posted my comment before I got sent a notification ;)